optimizely/event/user_event_factory.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2019 Optimizely
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .user_event import EventContext, ConversionEvent, ImpressionEvent
+from .event_factory import EventFactory
+
+
+class UserEventFactory(object):
+  """ UserEventFactory builds impression and conversion events from a given UserEvent. """
+
+  @classmethod
+  def create_impression_event(cls, project_config, activated_experiment, variation_id, user_id, user_attributes):
+    """ Create impression Event to be sent to the logging endpoint.
+
+    Args:
+      project_config: Instance of ProjectConfig.
+      experiment: Experiment for which impression needs to be recorded.
+      variation_id: ID for variation which would be presented to user.
+      user_id: ID for user.
+      attributes: Dict representing user attributes and values which need to be recorded.
+
+    Returns:
+      Event object encapsulating the impression event.
+    """
+
+    experiment_key = activated_experiment.key if activated_experiment else None
```

Comment:
The API requires experiment id to be sent in a decision. So I don't think we can create a valid impression event if `activated_experiment` is `None`.

Wouldn't it be better to return `None` from `create_impression_event` in this case? Otherwise we're creating an `ImpressionEvent` that won't be counted by the API.